File: pages/refactorings/replace-mutable-fields-with-lenses/02-player-position/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("moveX: Before == After", agreesX),
    property("moveY: Before == After", agreesY),
    property("the composed path lens player.x obeys the laws", laws),
  )

  // Raw values, so the same input feeds both Game types.
  val genInt: Gen[Int] = Gen.int(Range.linear(-30, 30))
  val genId: Gen[String] =
    for n <- Gen.int(Range.linear(0, 9)) yield "p" + n.toString

  def agreesX: Property =
    for
      id <- genId.forAll
      lv <- genInt.forAll
      px <- genInt.forAll
      py <- genInt.forAll
      dx <- genInt.forAll
    yield
      val b =
        Before.moveX(Before.Game(lv, Before.Player(id, px, py)), dx)
      val a = After.moveX(After.Game(lv, After.Player(id, px, py)), dx)
      (b.level, b.player.id, b.player.x, b.player.y)
        ==== (a.level, a.player.id, a.player.x, a.player.y)

  def agreesY: Property =
    for
      id <- genId.forAll
      lv <- genInt.forAll
      px <- genInt.forAll
      py <- genInt.forAll
      dy <- genInt.forAll
    yield
      val b =
        Before.moveY(Before.Game(lv, Before.Player(id, px, py)), dy)
      val a = After.moveY(After.Game(lv, After.Player(id, px, py)), dy)
      (b.level, b.player.id, b.player.x, b.player.y)
        ==== (a.level, a.player.id, a.player.x, a.player.y)

  def laws: Property =
    for
      id <- genId.forAll
      lv <- genInt.forAll
      px <- genInt.forAll
      py <- genInt.forAll
      v1 <- genInt.forAll
      v2 <- genInt.forAll
    yield
      val g  = After.Game(lv, After.Player(id, px, py))
      val l  = After.playerX
      l.get(l.set(v1)(g)) ==== v1 and
        l.set(l.get(g))(g) ==== g and
        l.set(v2)(l.set(v1)(g)) ==== l.set(v2)(g)

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default),
      t.result, Seed.fromTime())
    println(Test.renderReport(
      "Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)

File: pages/refactorings/replace-mutable-fields-with-lenses/02-player-position/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

genInt :: Gen Int
genInt = Gen.int (Range.linear (-30) 30)

genId :: Gen String
genId = ("p" <>) . show <$> Gen.int (Range.linear 0 9)

prop_agrees :: Property
prop_agrees = property $ do
  id <- forAll genId
  lv <- forAll genInt
  px <- forAll genInt
  py <- forAll genInt
  dx <- forAll genInt
  dy <- forAll genInt
  let bg = Before.Game lv (Before.Player id px py)
      ag = After.Game lv (After.Player id px py)
      b1 = Before.moveX bg dx
      b2 = Before.moveY bg dy
      a1 = After.moveX ag dx
      a2 = After.moveY ag dy
  (Before.level b1, Before.px (Before.gplayer b1),
   Before.py (Before.gplayer b1))
    === (After.level a1, After.px (After.gplayer a1),
         After.py (After.gplayer a1))
  (Before.level b2, Before.px (Before.gplayer b2),
   Before.py (Before.gplayer b2))
    === (After.level a2, After.px (After.gplayer a2),
         After.py (After.gplayer a2))

-- The composed path is still a lens: get and set through player . x
-- satisfy the three equations.
prop_laws :: Property
prop_laws = property $ do
  id <- forAll genId
  lv <- forAll genInt
  px <- forAll genInt
  py <- forAll genInt
  v1 <- forAll genInt
  v2 <- forAll genInt
  let l = After.playerX
      g = After.Game lv (After.Player id px py)
  After.set l (After.get l g) g === g
  After.get l (After.set l v1 g) === v1
  After.set l v2 (After.set l v1 g) === After.set l v2 g

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("moveX/moveY: Before == After", prop_agrees)
    , ("composed path lens player.x obeys the laws", prop_laws)
    ]
  unless ok exitFailure
